e2e: check regexp.MatchReader return, curl SSL issue

gyuho opened the pull request:
```go
buf := new(bytes.Buffer)
buf.WriteString("abc")
s, err := regexp.MatchReader("x", buf)
fmt.Println(s, err)
// false <nil>
```

1. proc.ExpectRegex returns the result of regexp.MatchReader,
which does not return error even if there is no match of regex.
This fixes it by checking the boolean value and if the boolean
value is false, it returns error.

1. Adds more tests and finishes https://github.com/coreos/etcd/issues/4259.

1. When we do the regex match correctly, curl request through SSL
returns error. For the purpose of debugging, I changed it to log
without failing the tests. etcdctl with SSL works fine [1]. 

1. Did some clean-ups.

1. Added TODO for watch: `// TODO: 'watch' sometimes times out in Semaphore CI environment but works fine in every other environments`.

1. increase test time

1:
```
--- PASS: TestCurlAllTLS (0.31s)
curl: (35) error:14094410:SSL routines:SSL3_READ_BYTES:sslv3 alert handshake failure
```

gyuho commented on e2e/process_test.go:
```diff
@@ -0,0 +1,466 @@
+	if err != nil {
+		return err
+	}
+	if !ok {
```
This is the part that I changed compared to previous `spawnWithExpect`.

heyitsanthony commented:
I'm fine with splitting out the generic utility stuff (e.g., curl, spawnWithExec), but please keep the etcd and etcdctl specific code in `etcd_test.go` and `etcdctl_test.go`; it would help avoid having one monolithic code file as the tests grow and the diff will be more readable.

heyitsanthony commented on e2e/process_test.go:
```diff
@@ -0,0 +1,469 @@
+	}
+	if epc.cfg.proxySize > 0 {
+		for i := 0; i < 5; i++ {
+			ok, _ := isProxyReady(epc)
```
the proxy is claiming it's "listening for client requests" even though it's not necessarily ready to accept client connections?

gyuho replied:
Yes, it prints that out before serving client requests: https://github.com/coreos/etcd/blob/master/etcdmain/etcd.go#L462-L466.

heyitsanthony commented on e2e/process_test.go:
```diff
@@ -0,0 +1,469 @@
+	if !ok {
+		return fmt.Errorf("couldn't get expected output: '%s'", expected)
+	}
+	return proc.Close()
```
I think `Close()` should be deferred once spawnCmd succeeds (the older version of the code should have done that but didn't)

heyitsanthony replied:
Nevermind the defer bit-- Close can't be deferred because the success case needs the close error code. The proc still needs a close on the error paths, though.

gyuho commented:
Agree. Will fix tonight. Think this made code review harder. Thanks.

On Sun, Jan 31, 2016, 14:09 Anthony Romano <notifications@github.com> wrote:

> I'm fine with splitting out the generic utility stuff (e.g., curl,
> spawnWithExec), but please keep the etcd and etcdctl specific code in
> etcd_test.go and etcdctl_test.go; it would help avoid having one
> monolithic code file as the tests grow and the diff will be more readable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4361#issuecomment-177631718>.
>
-- 
Sincerely,
Gyu-Ho Lee

gyuho commented:
@heyitsanthony PTAL. Split util functions into two files as it was. Thanks!

xiang90 commented:
@gyuho Seems fine with me. Defer to @heyitsanthony

heyitsanthony commented on e2e/etcd_test.go:
```diff
@@ -99,35 +105,37 @@ func testProcessClusterPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
 	expectPut := `{"action":"set","node":{"key":"/testKey","value":"foo","`
 	if err := cURLPut(epc, "testKey", "foo", expectPut); err != nil {
-		t.Fatalf("failed put with curl (%v)", err)
+		t.Logf("[WARNING] failed put with curl (%v)", err)
```
shouldn't the test fail if put fails?

gyuho replied:
It fails in Semaphore CI and my local machine, but I think it's a cert issue. Any idea?

```
curl: (35) error:14094410:SSL routines:SSL3_READ_BYTES:sslv3 alert handshake failure
```

Possibly related to https://github.com/coreos/etcd/issues/209

I will see if I can fix it.

Thanks

heyitsanthony commented on e2e/etcd_test.go:
```diff
@@ -198,9 +212,72 @@ func newEtcdProcessCluster(cfg *etcdProcessClusterConfig) (*etcdProcessCluster,
 	return epc, nil
 }
 
+func isProxyReady(clus *etcdProcessCluster) (bool, error) {
```
I don't understand why this is necessary-- isn't the proxy ready after it says it's listening?

heyitsanthony commented on e2e/etcd_test.go:
```diff
@@ -313,19 +390,10 @@ func (epc *etcdProcessCluster) Close() (err error) {
 	// redirect stderr to stdout since gexpect only uses stdout
-	cmd := `/bin/sh -c "` + strings.Join(args, " ") + ` 2>&1 "`
-	return gexpect.Spawn(cmd)
+	cmdArgs := `/bin/sh -c "` + strings.Join(args, " ") + ` 2>&1 "`
```
this is a full command, not only its arguments

gyuho replied:
Right. Will fix.

heyitsanthony commented on e2e/etcd_test.go:
```diff
@@ -333,8 +401,39 @@ func spawnWithExpect(args []string, expected string) error {
 		return err
 	}
-	if _, err := proc.ExpectRegex(expected); err != nil {
+	ok, err := proc.ExpectRegex(expected)
```
I think process should be cleaned up regardless of code path
```go
ok, err := proc.ExpectRegex(expected)
perr := proc.Close()
...
return perr
```

gyuho commented:
@heyitsanthony Addressed your feedback. PTAL.

And regarding `isProxyReady` function.

Current proxy implementation has its own discovery function https://github.com/coreos/etcd/blob/master/proxy/director.go#L64-L85 which can be delayed sometimes. 

And https://github.com/coreos/etcd/blob/master/etcdmain/etcd.go#L461-L467 shows that it logs `proxy: listening for client requests on` before the actual serving.

```go
		go func() {
			plog.Info("proxy: listening for client requests on ", host)
			mux := http.NewServeMux()
			mux.Handle("/metrics", prometheus.Handler())
			mux.Handle("/", ph)
			plog.Fatal(http.Serve(l, mux))
		}()
```

Without `isProxyReady`, it works half of the time, but other half fails because proxy hasn't found the endpoints. The test failed with `no endpoints available`. It's kind of hacky way. We need to better implementation in proxy side, I think.

gyuho commented:
And I created a separate issue for `curl` tests. https://github.com/coreos/etcd/issues/4368. I will fix the curl issue in a separate PR. Thanks

gyuho commented:
Merge conflict... will rebase.

gyuho commented:
nvm. No merge conflict. Misread. 3rd party integration disabled.

heyitsanthony commented:
@gyuho maybe we can have a log message for the proxy that announces it's discovered endpoints on initialization and wait on that instead of the `listening for client requests` string? /cc @xiang90

gyuho commented:
I changed some code in proxy package to print out endpoints when there are at lease one endpoint:

```
proxy: endpoints found ["http://127.0.0.1:12379" "http://127.0.0.1:22379" "http://127.0.0.1:32379"]
```

PTAL.

/cc @heyitsanthony @xiang90

xiang90 commented:
@heyitsanthony I am fine with adding that log line as long as it does not spam every time it updates (only print out when there is a change)

heyitsanthony commented on proxy/director.go:
```diff
@@ -38,12 +40,22 @@ func newDirector(urlsFunc GetProxyURLs, failureWait time.Duration, refreshInterv
+					for _, e := range es {
+						sl = append(sl, e.URL.String())
+					}
+					plog.Infof("endpoints found %q", sl)
```
`proxy:`?

gyuho replied:
Here https://github.com/coreos/etcd/blob/master/proxy/reverse.go#L37

heyitsanthony commented on proxy/director.go:
```diff
@@ -69,7 +81,7 @@ func (d *director) refresh() {
 		uu, err := url.Parse(u)
 		if err != nil {
-			log.Printf("proxy: upstream URL invalid: %v", err)
+			plog.Printf("upstream URL invalid: %v", err)
```
`proxy:`?

xiang90 replied:
we use plog(pkg logger) here, which attaches the pkg prefix.

gyuho replied:
`capnslog` prefixes it automatically.

heyitsanthony replied:
good to know, thanks

heyitsanthony commented:
lgtm but `testProcessClusterPutGet` should be fixed to use `Fatalf` soon

gyuho commented:
@heyitsanthony Thanks. Yeah I will fix it in a separate PR. Will merge when green lights.
